pkg/controllers/management/aks/aks_cluster_handler.go

```diff
@@ -0,0 +1,723 @@
+	}
+
+	apiEndpoint := string(caSecret.Data["endpoint"])
+	caCert := string(caSecret.Data["ca"])
+	if cluster.Status.APIEndpoint == apiEndpoint && cluster.Status.CACert == caCert {
+		return cluster, nil
+	}
+
+	var currentCluster *mgmtv3.Cluster
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		currentCluster, err = e.clusterClient.Get(cluster.Name, v1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		currentCluster.Status.APIEndpoint = apiEndpoint
+		currentCluster.Status.CACert = base64.StdEncoding.EncodeToString(caSecret.Data["ca"])
+		currentCluster, err = e.clusterClient.Update(currentCluster)
+		return err
+	})
+
+	return currentCluster, err
+}
+
+// generateAndSetServiceAccount uses the API endpoint and CA cert to generate a service account token. The token is then copied to the cluster status.
+func (e *aksOperatorController) generateAndSetServiceAccount(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+
+	restConfig, err := e.getKubeConfig(cluster)
+	if err != nil {
+		return cluster, err
+	}
+	saToken, err := generateSAToken(restConfig)
+	if err != nil {
+		return cluster, err
+	}
+
+	cluster = cluster.DeepCopy()
+	cluster.Status.ServiceAccountToken = saToken
+	return e.clusterClient.Update(cluster)
+}
+
+// buildAKSCCCreateObject returns an object that can be used with the kubernetes dynamic client to
+// create an AKSClusterConfig that matches the spec contained in the cluster's AKSConfig.
+func buildAKSCCCreateObject(cluster *mgmtv3.Cluster) (*unstructured.Unstructured, error) {
+	aksClusterConfig := aksv1.AKSClusterConfig{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "AKSClusterConfig",
+			APIVersion: aksV1,
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: cluster.Name,
+			OwnerReferences: []v1.OwnerReference{
+				{
+					Kind:       cluster.Kind,
+					APIVersion: rbac.RancherManagementAPIVersion,
+					Name:       cluster.Name,
+					UID:        cluster.UID,
+				},
+			},
+		},
+		Spec: *cluster.Spec.AKSConfig,
+	}
+
+	// convert AKS cluster config into unstructured object so it can be used with dynamic client
+	aksClusterConfigMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&aksClusterConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &unstructured.Unstructured{
+		Object: aksClusterConfigMap,
+	}, nil
+}
+
+// recordAppliedSpec sets the cluster's current spec as its appliedSpec
+func (e *aksOperatorController) recordAppliedSpec(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	if reflect.DeepEqual(cluster.Status.AppliedSpec.AKSConfig, cluster.Spec.AKSConfig) {
+		return cluster, nil
+	}
+
+	cluster = cluster.DeepCopy()
+	cluster.Status.AppliedSpec.AKSConfig = cluster.Spec.AKSConfig
+	return e.clusterClient.Update(cluster)
+}
+
+// deployAKSOperator looks for the rancher-aks-operator app in the cattle-system namespace, if not found it is deployed.
+// If it is found but is outdated, the latest version is installed.
+func (e *aksOperatorController) deployAKSOperator() error {
```

Comment:
I think this may also be logic that can be shared among the KEv2 integration controllers.

Reply:
+1: perhaps we can move such things to another package and export them.

Reply:
Yes, I agree it will be reworked